timelock-impl/src/main/java/com/palantir/atlasdb/timelock/AsyncTimelockResource.java

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright 2017 Palantir Technologies, Inc. All rights reserved.
+ *
+ * Licensed under the BSD-3 License (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://opensource.org/licenses/BSD-3-Clause
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.timelock;
+
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+
+import javax.ws.rs.Consumes;
+import javax.ws.rs.POST;
+import javax.ws.rs.Path;
+import javax.ws.rs.Produces;
+import javax.ws.rs.container.AsyncResponse;
+import javax.ws.rs.container.Suspended;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+
+import com.palantir.lock.v2.LockImmutableTimestampRequest;
+import com.palantir.lock.v2.LockImmutableTimestampResponse;
+import com.palantir.lock.v2.LockRequestV2;
+import com.palantir.lock.v2.LockTokenV2;
+import com.palantir.lock.v2.WaitForLocksRequest;
+import com.palantir.timestamp.TimestampRange;
+
+@Path("/timelock")
+@Consumes(MediaType.APPLICATION_JSON)
+@Produces(MediaType.APPLICATION_JSON)
+public class AsyncTimelockResource {
+
+    private final AsyncTimelockService timelock;
+
+    public AsyncTimelockResource(AsyncTimelockService timelock) {
+        this.timelock = timelock;
+    }
+
+    @POST
+    @Path("fresh-timestamp")
+    public long getFreshTimestamp() {
+        return timelock.getFreshTimestamp();
+    }
+
+    @POST
+    @Path("fresh-timestamps")
+    public TimestampRange getFreshTimestamps(int numTimestampsRequested) {
+        return timelock.getFreshTimestamps(numTimestampsRequested);
+    }
+
+    @POST
+    @Path("lock-immutable-timestamp")
+    public LockImmutableTimestampResponse lockImmutableTimestamp(LockImmutableTimestampRequest request) {
+        return timelock.lockImmutableTimestamp(request);
+    }
+
+    @POST
+    @Path("immutable-timestamp")
+    public long getImmutableTimestamp() {
+        return timelock.getImmutableTimestamp();
+    }
+
+    @POST
+    @Path("lock")
+    public void lock(@Suspended final AsyncResponse response, LockRequestV2 request) {
+        CompletableFuture<LockTokenV2> future = timelock.lock(request);
+        future.whenComplete((token, error) -> {
+            if (error != null) {
+                response.resume(error);
+            } else {
+                response.resume(token);
+            }
+        });
+    }
```

Comment:
I did a quick test to explore this. (Note this is http 1.1 behavior from integTests):
- Interestingly, it looks like the jetty idle timeout is ignored for suspended requests. A request can block for more than the idle timeout
- There is an option to set a timeout on the `AsyncResponse`; but it's infinite by default
- If we block for more than the client okhttp timeout (60s by default), then the request gets silently retried. However this is fine due to the de-duping of `requestId` on the server side; only downside is that lots of responses are returned at once
- If we block for more than 5x the client timeout, then it throws a `feign.RetryableException: timeout executing`
- It's possible to register a `ConnectionCallback` on `AsyncResponse` which tells you if the client has disconnected. I was not able to get this callback to be invoked by killing the request thread or waiting for it to timeout.

Will add a note to the burndown list to verify behavior on http/2.